File: go/0015M-三数之和.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-16 19:45:02
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

import "sort"

func threeSum(nums []int) [][]int {
	ans := [][]int{}
	sort.Ints(nums)
	for i, v := range nums {
		if v > 0 {
			break
		}
		if i > 0 && v == nums[i-1] {
			continue
		}
		left, right := i+1, len(nums)-1
		for left < right {
			sum := v + nums[left] + nums[right]
			if sum > 0 {
				right--
			} else if sum < 0 {
				left++
			} else {
				ans = append(ans, []int{v, nums[left], nums[right]})
				// 去重
				for left < right && nums[left] == nums[left+1] {
					left++
				}
				for left < right && nums[right] == nums[right-1] {
					right--
				}
				left++
				right--
			}
		}
	}
	return ans
}

File: kt/src/0015M-三数之和.kt
fun threeSum(nums: IntArray): List<List<Int>> {
    nums.sort()
    val ans: MutableList<List<Int>> = mutableListOf()
    for (i in nums.indices) {
        if (i > 0 && nums[i] == nums[i - 1]) continue
        var left = i + 1
        var right = nums.size - 1
        while (left < right) {
            val sum = nums[i] + nums[left] + nums[right]
            when {
                sum < 0 -> left++
                sum > 0 -> right--
                else -> {
                    ans.add(listOf(nums[i], nums[left], nums[right]))
                    while (left < right && nums[right] == nums[right - 1]) right--
                    while (left < right && nums[left] == nums[left + 1]) left++
                    left++
                    right--
                }
            }
        }
    }
    return ans
}
